.github/workflows/node.js.yml

```diff
@@ -0,0 +1,107 @@
+      - staging
+      # - master
+  workflow_dispatch:
+
+jobs:
+  PreDeployTesting:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Repository checkout
+        uses: actions/checkout@v3
+
+      - name: Node.js setup
+        uses: actions/setup-node@v3
+        with:
+          node-version: '16.x'
+
+      - name: 📦 Install dependencies
+        run: yarn install
+
+      - name: 💅 Prettier check
+        run: yarn prettier:check
+
+      - name: 💨 Lint check
+        run: yarn lint
+
+      - name: 🏗️ Build app
+        run: yarn ng build --prod
+
+  Deploy:
+    if: ${{ github.event_name == 'push' && (github.ref == 'refs/heads/master' || github.ref == 'refs/heads/staging')}}
+    runs-on: ubuntu-latest
+    environment:
+      name: ${{ (github.ref == 'refs/heads/master' && 'prod') || 'stage' }}
+    needs: PreDeployTesting
+    permissions:
+      id-token: write
+      contents: read
+
+    steps:
+      - name: Repository checkout
+        uses: actions/checkout@v3
+
+      - name: Install AWSCLI
+        run: pip install --user awscli
+
+      - name: Configure AWS credentials
+        uses: aws-actions/configure-aws-credentials@036a4a1ddf2c0e7a782dca6e083c6c53e5d90321
+        with:
+          aws-region: us-east-1
+          role-to-assume: ${{ secrets.AWS_IAM_ROLE_ARN }}
+
+      - name: Node.js setup
+        uses: actions/setup-node@v3
+        with:
+          node-version: '16.x'
+
+      - name: 📦 Install dependencies
+        run: yarn install
+
+      - name: 🏗️ Build app
+        run: yarn ng build --prod
+
+      - name: Copy embed.js
+        run: cp embed/embed.js dist/embed.js
+
+      - name: Copy example.html
+        run: cp embed/example.html dist/example.html
+
+      - name: Copy mobile directory
+        run: cp -r mobile/ dist/mobile/
+
+      - name: Update embed host (staging)
+        if: github.ref == 'refs/heads/staging'
+        run: sed -i "s%$appDomainPlaceholder%$appDomain%g" dist/embed.js
+        env:
+          appDomainPlaceholder: '{appDomain}'
+          appDomain: 'https://stage.knowgod.com'
+
+      - name: Deploy to s3 (staging)
+        if: github.ref == 'refs/heads/staging'
+        run: |
+          aws s3 sync dist/knowgod s3://$S3_BUCKET --acl 'public-read' --exclude '*.html' --cache-control 'public, max-age=31536000' --delete
+          aws s3 sync dist/knowgod s3://$S3_BUCKET --acl 'public-read' --include '*.html' --cache-control 'no-cache' --delete
+          aws s3 cp dist/embed.js s3://$S3_BUCKET --acl 'public-read' --cache-control 'no-cache'
+          aws s3 cp dist/example.html s3://$S3_BUCKET --acl 'public-read' --cache-control 'no-cache'
+          aws s3 sync dist/mobile s3://$S3_BUCKET --acl 'public-read' --cache-control 'no-cache' --content-type 'application/json'
+        env:
+          S3_BUCKET: ${{ secrets.AWS_S3_BUCKET }}
+
+      - name: Deploy prod
+        if: github.ref == 'refs/heads/master'
+        run: echo "Deploy code goes here."
```

Comment:
you shouldn't need separate logic to deploy to production, you should be using the same 2 steps that are used to deploy to stage

Reply:
@frett Once I switch the `Update embed host: env: appDomain` to use the bucket secret, that block can be used for both staging and prod deployment. But the "Deploy to s3" block needs different caching values between the two environments. Do you have a way to dynamically swap those? I agree it could improve the maintainability of the code.

Reply:
we should probably make the caching values the same, I wonder if that even matters though because our usually S3 website config involves using CloudFront as a caching layer, which would mean S3 cache times are meaningless

Reply:
yeah, we are using CloudFront, which would be doing our caching for us

Reply:
So do I need to specify caching values here at all? Should I take out all the `cache-control` options and values?

Reply:
let's use the cache values from production, I think it's just setting a different cache setting for index.html

Reply:
I did research and CloudFront by default utilizes the origin (S3 in this case) cache headers to determine how long it caches data.

If we are concerned about delay between merging a commit and seeing the results live on the stage environment we should probably trigger a cache invalidation in CloudFront after we upload everything to S3. But that will be outside the scope of this PR and could be a future optimization

Reply:
relevant CloudFront documentation: https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/Expiration.html